packages/astro/src/content/types-generator.ts

```diff
@@ -342,58 +343,68 @@ function normalizeConfigPath(from: string, to: string) {
 }
 
-const schemaCache = new Map<string, ZodSchema>();
+const schemaContextResultCache = new Map<string, { schema: ZodSchema; types: string }>();
 
-async function getContentLayerSchema<T extends keyof ContentConfig['collections']>(
+async function getSchemaContextResult<T extends keyof ContentConfig['collections']>(
 	collection: ContentConfig['collections'][T],
 	collectionKey: T,
-): Promise<ZodSchema | undefined> {
-	const cached = schemaCache.get(collectionKey);
+) {
+	const cached = schemaContextResultCache.get(collectionKey);
 	if (cached) {
 		return cached;
 	}
 
 	if (
 		collection?.type === CONTENT_LAYER_TYPE &&
 		typeof collection.loader === 'object' &&
-		collection.loader.schema
+		!collection.loader.schema &&
+		collection.loader.getSchemaContext
 	) {
-		let schema = collection.loader.schema;
-		if (typeof schema === 'function') {
-			schema = await schema();
-		}
-		if (schema) {
-			schemaCache.set(collectionKey, await schema);
-			return schema;
-		}
+		const result = await collection.loader.getSchemaContext();
+		schemaContextResultCache.set(collectionKey, result);
+		return result;
 	}
 }
 
+async function getContentLayerSchema<T extends keyof ContentConfig['collections']>(
+	collection: ContentConfig['collections'][T],
+	collectionKey: T,
+): Promise<ZodSchema | undefined> {
+	if (collection?.type !== CONTENT_LAYER_TYPE || typeof collection.loader === 'function') {
+		return;
+	}
+	if (collection.loader.schema) {
+		return collection.loader.schema;
+	}
+	const result = await getSchemaContextResult(collection, collectionKey);
+	return result?.schema;
+}
+
 async function typeForCollection<T extends keyof ContentConfig['collections']>(
 	collection: ContentConfig['collections'][T] | undefined,
 	collectionKey: T,
-): Promise<string> {
+): Promise<{ type: string; injectedType?: InjectedType }> {
 	if (collection?.schema) {
-		return `InferEntrySchema<${collectionKey}>`;
+		return { type: `InferEntrySchema<${collectionKey}>` };
 	}
-	if (!collection?.type) {
-		return 'any';
+	if (!collection?.type || typeof collection.loader === 'function') {
+		return { type: 'any' };
 	}
-	const schema = await getContentLayerSchema(collection, collectionKey);
-	if (!schema) {
-		return 'any';
+	if (collection.loader.schema) {
+		return { type: `InferLoaderSchema<${collectionKey}>` };
 	}
-	try {
-		const zodToTs = await import('zod-to-ts');
-		const ast = zodToTs.zodToTs(schema);
-		return zodToTs.printNode(ast.node);
-	} catch (err: any) {
-		// zod-to-ts is sad if we don't have TypeScript installed, but that's fine as we won't be needing types in that case
-		if (err.message.includes("Cannot find package 'typescript'")) {
-			return 'any';
-		}
-		throw err;
+	const result = await getSchemaContextResult(collection, collectionKey);
+	if (!result) {
+		return { type: 'any' };
 	}
+	const filename = `loaders/${collectionKey.slice(1, -1)}.ts`;
+	return {
+		type: `import("./${filename}").Collection`,
```

Comment:
This isn't the filename, is it? It should be the runtime type, so with a `.js` extension